### PR TITLE
* adds commit replace with config file for iosxr

### LINKED ISSUE #32061: iosxr_config src parameter does not do what is expected.
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - iosxr_config module

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

```

##### CONFIGURATION
No settings changed in ansible.cfg
N/A

##### OS / ENVIRONMENT
ubuntu trusty 14.04.5- from where i am running playbook
cisco iosxr:  Version 6.1.2 - router where i want to replace config

##### SUMMARY
The src parameter in iosxr_config module is not working as expected.As mentioned in the example in the module guide:
```yaml
- name: load a config from disk and replace the current config
  iosxr_config:
    src: config.cfg
    backup: yes
```
I am trying to replace the iosxr_config with simple the configuration stored in file(running-config).it does not do "commit replace", instead it merges the configuration. if i use replace paramete like below:
```yaml
- name: load a config from disk and replace the current config
  iosxr_config:
    src: config.cfg
    backup: yes
    replace: config
```
It does not work and throws error.

##### STEPS TO REPRODUCE

set_config.yml

```yaml
---
- hosts: all
  remote_user: cisco
  #become: true
  gather_facts: no
  tasks:
   - name: replacing the iosxr configuration
     iosxr_config:
       src: running-config
       backup: yes
       replace: config
```
ansible_host config:
```yaml
Node_1   ansible_host=172.16.1.100 ansible_port=22 ansible_connection=local ansible_ssh_user=admin ansible_ssh_pass=admin
```
configuration to be replaced(stored in file running-config):

Building configuration...
!! IOS XR Configuration 6.1.2
!! Last configuration change at Tue Oct 24 05:17:00 2017 by cisco
!
hostname Node_1
interface MgmtEth0/0/CPU0/0
 description OOB Management
 ! Configured on launch
 ipv4 address 172.16.1.127 255.255.255.0
!
ssh server v2
ssh server vrf default
ssh server logging
end

##### EXPECTED RESULTS

should replace the configuration on the router(equivalent to "commit replace")

##### ACTUAL RESULTS

```java

$ ansible-playbook -i ansible_hosts set_config.yml 

PLAY [all] ******************************************************************************************************************************************************************

TASK [replacing the iosxr configuration] ************************************************************************************************************************************
fatal: [Node_1]: FAILED! => {"changed": false, "commands": ["Tue Oct 24 05:17:07.664 UTC", "Building configuration...", "hostname Node_1", "interface MgmtEth0/0/CPU0/0", "description OOB Management", "ipv4 address 172.16.1.127 255.255.255.0", "ssh server v2", "ssh server vrf default", "ssh server logging", "end"], "failed": true, "msg": "Tue Oct 24 05:17:07.664 UTC\r\n\r                            ^\r\n% Invalid input detected at '^' marker.\r\nRP/0/0/CPU0:Node_1(config)#", "rc": 1}
	to retry, use: --limit @/home/cisco/gaurav/set_config.retry

PLAY RECAP ******************************************************************************************************************************************************************
Node_1                     : ok=0    changed=0    unreachable=0    failed=1

```

### PR DESCRIPTION
**Pull request by kedarkekan**
* fixes dci failure in iosxr_logging

##### SUMMARY

fixes #32061 

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
iosxr_config
iosxr_logging
##### ANSIBLE VERSION

```
2.5.0
```

### CODE REVIEW

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 1]
```diff
@@ -415,7 +416,14 @@ def load_config(module, command_filter, commit=False, replace=False,
         if module._diff:
             diff = get_config_diff(module)
 
-        if commit:
+        if replace:
+            cmd = {'command': 'commit replace',
+                   'prompt': 'This commit will replace or remove the entire running configuration',
+                   'answer': 'yes'}
+            conn.edit_config(cmd)
+            sleep(5)
```
Comment (gundalow) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: Things like this scary me.

Comment (kedarkekan) [reply lib/ansible/module_utils/network/iosxr/iosxr.py]: So does me, usually the commit replace operation takes say like a 1 or 2 second to complete. I am not sure if our command are sent async; assuming they are async, I added a delay here. With a small config file it works just fine without delay too. Maybe I will comment this out for now, with a note stating add a delay if a failure occurs in future (which I hope won't). BTW this delay is only during replace config operation

Comment (ganeshrn) [reply lib/ansible/module_utils/network/iosxr/iosxr.py]: afaik after sending command to remote host, we wait for response till default command timeout is triggered (30 sec)

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 2]
```diff
@@ -428,20 +436,36 @@ def run_command(module, commands):
...
         try:
             responses.append(to_text(out, errors='surrogate_or_strict'))
         except UnicodeError:
             module.fail_json(msg=u'failed to decode output from {0}:{1}'.format(cmd, to_text(out)))
     return responses
+
+
+def put_file(module, src, dst, proto='scp'):
+    conn = get_connection(module)
+    conn.put_file_cliconf(source=src, destination=dst, proto=proto)
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: Can we rename this to `copy_file` and `get_file`. Can't think of a reason why API name should have connection type in it.

File: lib/ansible/plugins/cliconf/__init__.py [hunk 3]
```diff
@@ -181,18 +180,26 @@ def discard_changes(self):
...
-            scp.put(source, destination)
-
-    def fetch_file(self, source, destination):
-        """Fetch file over scp from remote device"""
-        if not HAS_SCP:
-            self._connection.internal_error("Required library scp is not installed.  Please install it using `pip install scp`")
-        ssh = self._connection._connect_uncached()
-        with SCPClient(ssh.get_transport()) as scp:
-            scp.get(source, destination)
+    def put_file_cliconf(self, source=None, destination=None, proto='scp'):
```
Comment (ganeshrn) [inline lib/ansible/plugins/cliconf/__init__.py]: Related to above comment on API name

File: lib/ansible/plugins/connection/network_cli.py [hunk 4]
```diff
@@ -288,6 +288,8 @@ def _connect(self):
         p.force_persistence = self.force_persistence
         ssh = p._connect()
 
+        self.p = p
```
Comment (ganeshrn) [inline lib/ansible/plugins/connection/network_cli.py]: `p` can be removed completely and `self.p` can be renamed to a meaningful name something like`self.paramiko_conn`

File: lib/ansible/module_utils/network/iosxr/iosxr.py [hunk 5]
```diff
@@ -415,7 +416,15 @@ def load_config(module, command_filter, commit=False, replace=False,
...
 
-        if commit:
+        if replace:
+            cmd = {'command': 'commit replace',
+                   'prompt': 'This commit will replace or remove the entire running configuration',
+                   'answer': 'yes'}
+            conn.edit_config(cmd)
+            # TODO: add a delay if commit replace operation takes too long
+            # sleep(5)
+            conn.edit_config('end')
```
Comment (ganeshrn) [inline lib/ansible/module_utils/network/iosxr/iosxr.py]: cmd can be a list and commands can be sent as a list to avoid invoking `edit_config` twice`?

**Review by gundalow — APPROVED** [review]

**Review by ganeshrn — APPROVED** [review]

**Review by trishnaguha — APPROVED** [review]

**Pull request merged**